mm2src/coins/utxo/utxo_common.rs

```diff
@@ -1353,31 +1436,55 @@ where
     T: AsRef<UtxoCoinFields> + UtxoCommonOps + MarketCoinOps,
 {
     let decimals = coin.as_ref().decimals;
-    let to = coin
-        .address_from_str(&req.to)
-        .map_to_mm(WithdrawError::InvalidAddress)?;
 
     let conf = &coin.as_ref().conf;
-    let is_p2pkh = to.prefix == conf.pub_addr_prefix && to.t_addr_prefix == conf.pub_t_addr_prefix;
-    let is_p2sh = to.prefix == conf.p2sh_addr_prefix && to.t_addr_prefix == conf.p2sh_t_addr_prefix && conf.segwit;
 
-    let script_pubkey = if is_p2pkh {
-        Builder::build_p2pkh(&to.hash)
-    } else if is_p2sh {
-        Builder::build_p2sh(&to.hash)
+    let script_pubkey = if let Ok(to) = SegwitAddress::from_str(&req.to) {
+        if Some(to.hrp.clone()) != conf.bech32_hrp {
+            let error = format!("Address {} is not a valid bech32 address for {}", to, conf.ticker);
+            MmError::err(WithdrawError::InvalidAddress(error))
+        } else {
+            Ok(Builder::build_p2wpkh(&AddressHash::from(&to.program[..])))
+        }
     } else {
-        let error = "Expected either P2PKH or P2SH".to_owned();
-        return MmError::err(WithdrawError::InvalidAddress(error));
-    };
+        let to = if conf.default_address_format.is_cashaddress() {
```

Comment:
This code does not validate that a valid cash address prefix is used, so I can pass `bitcoincash:qqyf96yqdrpa8f6pkf9f00ap068m5tgvly28qsfq9p` to the `mm2_tests::test_withdraw_cashaddresses` test and it will generate a transaction instead of returning the error.
Please add a test for this and also simplify the code - it already parses the address e.g. `CashAddress::from_str` but then parses it again for no reason in `address_from_str_and_format`.

Reply:
`CashAddress::from_str` returns `CashAddress` Struct but `address_from_str_and_format`  returns `Address` Struct which what's needed here that's why I parse it again. I use `CashAddress::from_str` to check if the withdraw `to` address is a valid cashaddress if not then I check if it's legacy. There is lot's of checks here but all is necessary and I can't find a way to remove some of it without a huge refactoring.